### PR TITLE
Refactor the propagation integration tests

### LINKED ISSUE #21: Debugging the propagation integration test is painful
[GitLab issue: #687](https://gitlab.cern.ch/acts/acts-core/issues/687)

The `PropagationIntegrationTest` does a lot of work. Its execution time is somewhat reasonable in "release" `-O2` builds, at around a minute here, but that degrades into a painful extreme of 2 hours and a half when run on "debug" `-O0` builds, which if the usual 30x orders of magnitude apply means that it would typically take several days to run such a build under a valgrind tool (memcheck, verrou...).

I personally find this a bit concerning, because some bugs are hard to track down on a release builds, even with debugging symbols on. But maybe I'm overreacting an others don't think that this is a problem?

If we do agree that this is a problem, how do we want to approach it? Here are some tentative ideas:

- Break the test down into multiple sub-programs.
- Evaluate wisdom of tuning down iteration count in long-running parts.
- If that loses coverage/physics, then consider allowing the caller of the test to execute or replay only a subset of it (e.g. "re-run iterations 10000 to 10100").

### PR DESCRIPTION
**Pull request by msmk0**
This PR splits the existing single propagation integration tests into smaller ones that each only test a single propagator type. Comparisons between different propagator types are also implemented as separate tests. The test code itself has also been refactor to reduce duplication.

These changes should go in before #327 and #328 in order to simplify finding/debugging errors that showed up during the development.

Fixes #21 (hopefully).

### CODE REVIEW

**Comment by msmk0** [comment]
@FabianKlimpel I think I have converted all tests and this is ready to be reviewed. There are a few new failures that show up and I disabled those tests for now. They occur probably because I test properties and results that were not tested in the previous version. I will open bug reports where possible.

**Comment by msmk0** [comment]
@FabianKlimpel I managed to fix the dense propagation tests and they are now included again in the integration tests. I reordered `PropagationTests.hpp` a bit to better group functions with similar functionality. A small README in the `Tests/IntegrationTests` explains the overall setup.

Please have a look as soon as you can.

File: Tests/IntegrationTests/PropagationAtlasConstant.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      makePropagator(bz), makeRiddersPropagator(bz), geoCtx, magCtx,
+      makeParametersCurvilinearWithCovariance(phi, theta, p, q), s,
+      ZCylinderSurfaceBuilder(), epsPos, epsDir, epsMom, epsCov, showDebug);
+}
+
+// NOTE msmk 2020-08-13
+// i suspect that this does not work because the Ridders propagator does not now
+// about the parameter bounds, e.g. that the radial component is always
+// positive. there are also some NaNs in the covariance from the regular
+// propagator.
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationAtlasConstant.cpp:140]: No, I did not. However, I do not think that we should specifically test the corner cases like this one. These are the kind of things that people will spend days over wondering what is going on. I would also prefer to keep the commented code in the file as a clear reminder that there is something not working or not quite understood. If we remove this, it will get lost.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+// absolute parameter tolerances for position, direction, and absolute momentum
+constexpr auto epsPos = 1_um;
+constexpr auto epsDir = 0.125_mrad;
+constexpr auto epsMom = 1_eV;
+// relative covariance tolerance
+constexpr auto epsCov = 0.0125;
+constexpr bool showDebug = false;
+
+constexpr auto bz = 2_T;
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp]: Probably laziness. It now varies as for all in the single propagator tests.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+}  // namespace
+
+BOOST_AUTO_TEST_SUITE(PropagationCompareAtlasEigenConstant)
+
+BOOST_DATA_TEST_CASE(
+    Forward,
+    ds::phi* ds::theta* ds::absMomentum* ds::chargeNonZero* ds::pathLength, phi,
+    theta, p, q, s) {
+  runFreePropagationComparisonTest(
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp:62]: It tests the free propagation in contrast to the bound propagation onto a surface.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+BOOST_DATA_TEST_CASE(
+    ToPlane,
+    ds::phi* ds::theta* ds::absMomentum* ds::chargeNonZero* ds::pathLength, phi,
+    theta, p, q, s) {
+  runToSurfaceComparisonTest(atlasPropagator, eigenPropagator, geoCtx, magCtx,
+                             makeParametersCurvilinear(phi, theta, p, q), s,
+                             PlaneSurfaceBuilder(), epsPos, epsDir, epsMom,
+                             epsCov, showDebug);
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp]: That is because it is arbitrary (see laziness above). I added tests for all surface types as for all the other tests.

The disc surface tests currently fail because the resulting phi coordinate on the disc is different for the ATLAS and Eigen propagator. This might be related to the problem discussed above: because the expected position on the disc is at the center, the local r vanishes and the phi coordinate becomes meaningless.

I left the tests commented for the reasons outlined above.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+    CHECK_CLOSE_COVARIANCE(cmp.covariance().value(), ref.covariance().value(),
+                           relativeTolerance);
+  }
+}
+
+// helpers to construct target surfaces from track states
+
+/// Construct the transformation from the curvilinear to the global coordinates.
+template <typename charge_t>
+inline std::shared_ptr<Acts::Transform3D> makeCurvilinearTransform(
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:167]: Because this can take arbitrary bound parameters as input, but I want the constructed surface to always be perpendicular to the track direction. Using the curvilinear transform is the simplest way to get this because it automatically defines the two local coordinates.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+#include "Acts/Surfaces/DiscSurface.hpp"
+#include "Acts/Surfaces/PlaneSurface.hpp"
+#include "Acts/Surfaces/StrawSurface.hpp"
+#include "Acts/Tests/CommonHelpers/FloatComparisons.hpp"
+#include "Acts/Utilities/UnitVectors.hpp"
+#include "Acts/Utilities/detail/periodic.hpp"
+
+#include <utility>
+
+// parameter construction helpers
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:24]: My plan is to improve the parameter constructors so that this helper functions becomes obsolete. This will happen in about one or two PRs after this one.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+  fwdOptions.debug = showDebug;
+  // backward propagation
+  options_t<Actions, Aborts> bwdOptions(geoCtx, magCtx, getDummyLogger());
+  bwdOptions.direction = Acts::backward;
+  bwdOptions.pathLimit = -pathLength;
+  bwdOptions.maxStepSize = 1_cm;
+  bwdOptions.debug = showDebug;
+
+  // propagate parameters forward
+  auto fwdResult = propagator.propagate(initialParams, fwdOptions);
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp]: Great suggestion. I updated the tests.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+/// \warning Does not check that they are defined on the same surface.
+template <typename charge_t>
+inline void checkParametersConsistency(
+    const Acts::SingleTrackParameters<charge_t>& cmp,
+    const Acts::SingleTrackParameters<charge_t>& ref, double epsPos,
+    double epsDir, double epsMom) {
+  using namespace Acts;
+
+  // check stored parameters
+  CHECK_CLOSE_ABS(cmp.template get<eBoundLoc0>(),
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:125]: It does a bit. Not having this as a single function would introduce quite a bit of code duplication and decided to prefer less code duplication here. In all existing cases, the comparison function is called only once in a test and there is thus no ambiguity as to its origin.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+}
+
+/// Propagate the initial parameters once for the given path length and
+/// use the propagated parameters to define a target surface. Propagate the
+/// initial parameters again to the target surface. Verify that the surface has
+/// been found and the parameters are consistent.
+template <typename propagator_t, typename charge_t, typename surface_builder_t,
+          template <typename, typename>
+          class options_t = Acts::PropagatorOptions>
+inline void runToSurfaceTest(
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:340]: That function already does use `transportFreely` and `transportToSurface`. Or are you refering to something else here?

**Comment by msmk0** [comment]
> In general the refactoring is good idea.
> I have a few comments, mostly small ones. If applicable to various places, I just wrote them once.
> However I am very concerned about the commented code.

Thanks @FabianKlimpel for the thorough review. I hope, I could address all your points. Regarding the failing/commented tests: fixing the underlying issues is beyond the scope of this PR. We can merge this PR anyways as it just surfaces potential existing problems and fix them in subsequent PRs.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp:62]: The propagation is also bound to curvilinear surface, that's why I asked.

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:167]: Why would you use a perpendicular transformation for a given surface?

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:125]: As far as I can see it is called multiple times

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:340]: I'm not sure anymore what I meant here.

**Comment by FabianKlimpel** [comment]
I still have the big issue that currently the tests run through without any problems. Even in #86 there are no problems although the granularity is increased. So I would like to be sure that the issues don't have their origin in these changes and this I cannot see right now. Could you give more details about when the errors occur?

**Comment by msmk0** [comment]
> I still have the big issue that currently the tests run through without any problems. Even in #86 there are no problems although the granularity is increased. So I would like to be sure that the issues don't have their origin in these changes and this I cannot see right now. Could you give more details about when the errors occur?

I do not remember the specific errors. You can uncomment the relevant tests and run them. That should tell you exactly where the errors occur.

File: Tests/IntegrationTests/PropagationCompareEigenStraightLine.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+const EigenPropagator eigenPropagatorNonZero{EigenStepper(magFieldNonZero)};
+const StraightLinePropagator straightPropagator{StraightLineStepper()};
+
+}  // namespace
+
+BOOST_AUTO_TEST_SUITE(PropagationCompareEigenStraightLine)
+
+// TODO both the eigen stepper and the straight line stepper do not seem to
+//      handle the neutral parameters correctly. the results contain inf/nan.
+//      fix the propagators and re-enable the tests.
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareEigenStraightLine.cpp:59]: You can uncomment the tests to see the errors. Previously, there were no tests with neutral parameters. The code existed to generated them, but the test datasets only contained +-e as values.

P.S. I just saw this comment. I was apparently hidden in the conversation page.

**Comment by msmk0** [comment]
@FabianKlimpel I looked again at the disc-related tests. In the end, I managed to get them working by moving the origin of the generated disc surface away from the particle position. Its local position is thus far away from the rho=0 singularity. I also checked the previous tests; there the target disc surface was always slightly disturbed with respect to the curvilinear frame and we probably never tested the specific situation that arose here.

What I think happened here was, that at the dead center of the disc the phi coordinate is undefined and you only see positive changes for the rho coordinate. I think that confuses the Ridders propagator. Unfortunately, I also saw NaNs in the covariance from the regular propagator. The behavior at the singularity should be looked at in the future (but not in this PR).

That leaves only the tests with the neutral track parameters. This setup was not tested before and having it commented out in this PR does not reduces the covered test phase space. I suspect that it is a problem with the interpretation of the `q/p` parameter; I already have PRs in the pipeline for that.

Please have a look again and see if this satisfies your objections. This PR is blocking me and should not wait any longer.

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:126]: I might not have been clear: it is only called once in any given test. If a test fails, the test framework tells you which iteration it originated from and there is no ambiguity as to where it was called from.

File: Tests/IntegrationTests/PropagationAtlasConstant.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationAtlasConstant.cpp:140]: This is changed in the latest iteration.

File: Tests/IntegrationTests/PropagationCompareEigenStraightLine.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    ChargedZeroMagneticField,
+    ds::phi* ds::theta* ds::absMomentum* ds::chargeNonZero* ds::pathLength, phi,
+    theta, p, q, s) {
+  runFreePropagationComparisonTest(
+      eigenPropagatorZero, straightPropagator, geoCtx, magCtx,
+      makeParametersCurvilinear(phi, theta, p, q), s, epsPos, epsDir, epsMom,
+      epsCov, showDebug);
+}
+
+// TODO charged particle w/ infinite momentum and non-zero magnetic field
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareEigenStraightLine.cpp]: I extend to comment to clarify which tests are still missing.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+  BoundVector stddev = BoundVector::Zero();
+  // TODO use momentum-dependent resolutions
+  stddev[eBoundLoc0] = 15_um;
+  stddev[eBoundLoc1] = 80_um;
+  stddev[eBoundTime] = 25_ns;
+  stddev[eBoundPhi] = 1_degree;
+  stddev[eBoundTheta] = 1.5_degree;
+  stddev[eBoundQOverP] = 1_e / 10_GeV;
+  BoundSymMatrix corr = BoundSymMatrix::Identity();
+  corr(eBoundLoc0, eBoundLoc1) = corr(eBoundLoc1, eBoundLoc0) = 0.125;
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:71]: Unfortunately, no: Eigen does not define a dedicated symmetric matrix type. The `ActsSymMatrix...` types are regular matrices underneath.

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:168]: I want a surface for which I know that the track can intersect it. The easiest way to ensure that, is to use a surface that is perpendicular to the track direction at the point of intersection. because there distance to the surface changes maximally with the propagation length. That however defines only a normal direction and I need two additional unit vectors to construct the local system.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+                           relativeTolerance);
+  }
+}
+
+// helpers to construct target surfaces from track states
+
+/// Construct the transformation from the curvilinear to the global coordinates.
+template <typename charge_t>
+inline std::shared_ptr<Acts::Transform3D> makeCurvilinearTransform(
+    const Acts::SingleTrackParameters<charge_t>& params) {
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:169]: To be more specific about the allowed inputs. I do not need to support generic parameters here.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+  BOOST_CHECK_EQUAL(params.charge(), 0);
+
+  return params;
+}
+
+// helpers to compare track parameters
+
+/// Check that two parameters object are consistent within the tolerances.
+///
+/// \warning Does not check that they are defined on the same surface.
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:117]: I do not think I can. E.g. if I compare two curvilinear parameters they might have exactly the same parameters, global positions, etc., but not be defined on an identical surface because they originate from two separate propagation calls.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp:62]: The curvilinear surface is special: it is not supplied externally and only created after the propagation with a freely chosen propagation length. Compare this to the *true* bound propagation to an externally supplied surface: the propagation length is not a free parameter.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+  double time = 0.0;
+  Vector3D mom = absMom * makeDirectionUnitFromPhiTheta(phi, theta);
+  CurvilinearParameters params(std::nullopt, pos, mom, charge, time);
+
+  // ensure initial parameters are valid
+  CHECK_CLOSE_ABS(params.position(), pos, 0.125_um);
+  CHECK_CLOSE_ABS(params.time(), time, 1_ps);
+  CHECK_CLOSE_ABS(params.momentum(), mom, 0.125_eV);
+  // charge should be identical not just similar
+  BOOST_CHECK_EQUAL(params.charge(), charge);
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:51]: I think, I added this because I saw some weird NaNs in #327 and wanted to make sure that the parameters are actually well-defined before I start the propagation.

It does not hurt either and I would just leave it in for now.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+  CHECK_CLOSE_ABS(params.momentum(), mom, 0.125_eV);
+  // charge should be identical not just similar
+  BOOST_CHECK_EQUAL(params.charge(), 0);
+
+  return params;
+}
+
+// helpers to compare track parameters
+
+/// Check that two parameters object are consistent within the tolerances.
```
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:115]: This should stay as-is. The classes are named e.g. `SingleBoundTrackParameters` as the object contains multiple track parameters and not just one. So they are *parameters* objects.

**Comment by msmk0** [comment]
@FabianKlimpel I also found and replied your additional comments that where hidden before.

File: Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp:62]: Of course it's a special case but a free propagation sounds to me like something with free parameters like in #86

[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationCompareAtlasEigenConstant.cpp]: I changed the name to `runForwardComparisonTest`. This is consistent with the naming of the single propagator tests.

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:117]: Right, that could actually happen

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:169]: In that case I need to change it afterwards ;)

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:126]: As far as I remember the error message still points to this particular line, right?

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/IntegrationTests/PropagationTests.hpp:168]: Ok, got it. The name is just quite misleading if you read it with a certain assumption of what it should do.

**Comment by FabianKlimpel** [comment]
> @FabianKlimpel I looked again at the disc-related tests. In the end, I managed to get them working by moving the origin of the generated disc surface away from the particle position. Its local position is thus far away from the rho=0 singularity. I also checked the previous tests; there the target disc surface was always slightly disturbed with respect to the curvilinear frame and we probably never tested the specific situation that arose here.
> 
> What I think happened here was, that at the dead center of the disc the phi coordinate is undefined and you only see positive changes for the rho coordinate. I think that confuses the Ridders propagator. Unfortunately, I also saw NaNs in the covariance from the regular propagator. The behavior at the singularity should be looked at in the future (but not in this PR).
> 
> That leaves only the tests with the neutral track parameters. This setup was not tested before and having it commented out in this PR does not reduces the covered test phase space. I suspect that it is a problem with the interpretation of the `q/p` parameter; I already have PRs in the pipeline for that.
> 
> Please have a look again and see if this satisfies your objections. This PR is blocking me and should not wait any longer.

The Ridders is quite not aware of the unterdlying geometry and therefore the disc is a huge problem for it. Great thing though that you were able to fix it. I went through the combinations in #86 and saw that I didn't cover all of them but had quite a variety of different tests involving neutral parameters. So I assume that these issues will solve themselves.

Could you resolve the discussion that are resolved?

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Tests/IntegrationTests/PropagationTests.hpp:126]: Yes, but you still get the additional context.

**Comment by msmk0** [comment]
I resolved most discussions. There is only one left open regarding the higher covariance deviations for straw surfaces. I think we can leave that for the future.
